#include "flightdetaildialog.h"
#include "ui_flightdetaildialog.h"
#include "bookingdialog.h"
#include "../Common/protocol.h"
#include <QMessageBox>
#include <QDebug>
#include <QTableWidgetItem>
#include <QPushButton>
#include <QHeaderView>  // æ·»åŠ è¿™ä¸ªå¤´æ–‡ä»¶
#include <QJsonArray>   // æ·»åŠ è¿™ä¸ªå¤´æ–‡ä»¶
#include <QJsonDocument> // æ·»åŠ è¿™ä¸ªå¤´æ–‡ä»¶

FlightDetailDialog::FlightDetailDialog(const Flight &flight,
                                       ClientNetworkManager *networkManager,
                                       const QString &username,
                                       QWidget *parent)
    : QDialog(parent), ui(new Ui::FlightDetailDialog),
    flight(flight), networkManager(networkManager), currentUsername(username)
{
    ui->setupUi(this);
    setupUI();

    // è¿æ¥ä¿¡å·
    connect(ui->closeButton, &QPushButton::clicked, this, &FlightDetailDialog::onCloseButtonClicked);
    connect(networkManager, &ClientNetworkManager::messageReceived,
            this, &FlightDetailDialog::onMessageReceived);

    // åŠ è½½èˆ±ä½æ•°æ®
    loadCabinData();
}

FlightDetailDialog::~FlightDetailDialog()
{
    delete ui;
}

void FlightDetailDialog::setupUI()
{
    // è®¾ç½®çª—å£æ ‡é¢˜
    setWindowTitle(QString("èˆªç­è¯¦æƒ… - %1").arg(flight.getFlightNumber()));

    // æ˜¾ç¤ºèˆªç­ä¿¡æ¯
    ui->flightNumberLabel->setText(flight.getFlightNumber());
    ui->airlineLabel->setText(flight.getAirline());


    // åˆ¤æ–­é£æœºå‹å·å¹¶æ·»åŠ åˆ†ç±»
    QString aircraftType = flight.getAircraftType();
    QString aircraftDisplay = aircraftType;

    if (aircraftType == "A330" || aircraftType == "B787" || aircraftType == "B777") {
        aircraftDisplay += "ï¼ˆå®½ï¼‰";  // å®½ä½“æœº
    } else if (aircraftType == "A320" || aircraftType == "A319" ||
               aircraftType == "B737" || aircraftType == "ARJ21") {
        aircraftDisplay += "ï¼ˆçª„ï¼‰";  // çª„ä½“æœº
    } else {
        aircraftDisplay += "ï¼ˆä¸­ï¼‰";  // ä¸­å‹æœºæˆ–å…¶ä»–
    }

    ui->aircraftTypeLabel->setText(aircraftDisplay);

    // æ˜¾ç¤ºèˆªçº¿ä¿¡æ¯
    ui->departureCityLabel->setText(flight.getDepartureCity());
    ui->departureTimeLabel->setText(flight.getDepartureTime().toString("hh:mm"));
    ui->arrivalCityLabel->setText(flight.getArrivalCity());
    ui->arrivalTimeLabel->setText(flight.getArrivalTime().toString("hh:mm"));
    ui->durationLabel->setText(flight.getDurationString());

    // è®¾ç½®èˆªç­ä¿¡æ¯æ–‡æœ¬
    QString infoText = QString(
                           "âœˆï¸ èˆªç­ä¿¡æ¯\n"
                           "èˆªç­å·: %1\n"
                           "èˆªç©ºå…¬å¸: %2\n"
                           "æœºå‹: %3\n\n"
                           "ğŸ›« å‡ºå‘ä¿¡æ¯\n"
                           "åŸå¸‚: %4\n"
                           "æ—¶é—´: %5\n"
                           "æ—¥æœŸ: %6\n\n"
                           "ğŸ›¬ åˆ°è¾¾ä¿¡æ¯\n"
                           "åŸå¸‚: %7\n"
                           "æ—¶é—´: %8\n"
                           "æ—¥æœŸ: %9\n\n"
                           "â±ï¸ å…¶ä»–ä¿¡æ¯\n"
                           "é£è¡Œæ—¶é•¿: %10\n"
                           "çŠ¶æ€: %11\n"
                           "å‰©ä½™åº§ä½: %12ä¸ª"
                           ).arg(flight.getFlightNumber())
                           .arg(flight.getAirline())
                           .arg(aircraftDisplay)  // ä½¿ç”¨å¸¦åˆ†ç±»çš„æœºå‹
                           .arg(flight.getDepartureCity())
                           .arg(flight.getDepartureTime().toString("hh:mm"))
                           .arg(flight.getDepartureTime().toString("yyyy-MM-dd"))
                           .arg(flight.getArrivalCity())
                           .arg(flight.getArrivalTime().toString("hh:mm"))
                           .arg(flight.getArrivalTime().toString("yyyy-MM-dd"))
                           .arg(flight.getDurationString())
                           .arg(flight.getStatus())
                           .arg(flight.getAvailableSeats());

    ui->flightInfoText->setText(infoText);
}

void FlightDetailDialog::loadCabinData()
{
    if (!networkManager || !networkManager->isConnected()) {
        QMessageBox::warning(this, "é”™è¯¯", "æœªè¿æ¥åˆ°æœåŠ¡å™¨");
        return;
    }

    // å‘é€èˆ±ä½æŸ¥è¯¢è¯·æ±‚
    NetworkMessage msg;
    msg.type = CABIN_SEARCH_REQUEST;
    msg.data["flight_id"] = flight.getId();

    networkManager->sendMessage(msg);
    qDebug() << "å‘é€èˆ±ä½æŸ¥è¯¢è¯·æ±‚ï¼Œèˆªç­ID:" << flight.getId();
}

void FlightDetailDialog::displayCabinData(const QList<Cabin> &cabins)
{
    currentCabins = cabins;
    ui->cabinTable->setRowCount(cabins.size());

    // å…ˆè®¾ç½®è¡¨æ ¼åˆ—å®½å’Œè¡Œé«˜ - å¢å¤§æ•´ä½“å°ºå¯¸
    ui->cabinTable->horizontalHeader()->setStretchLastSection(false);

    // å¢å¤§æ‰€æœ‰åˆ—å®½ï¼Œè®©è¡¨æ ¼çœ‹èµ·æ¥æ›´èˆ’é€‚
    ui->cabinTable->setColumnWidth(0, 100);   // èˆ±ä½ç±»å‹ - åŠ å®½
    ui->cabinTable->setColumnWidth(1, 110);   // ä»·æ ¼ - åŠ å®½
    ui->cabinTable->setColumnWidth(2, 100);   // è¡Œæé¢åº¦ - åŠ å®½
    ui->cabinTable->setColumnWidth(3, 120);   // å¯ç”¨åº§ä½ - åŠ å®½
    ui->cabinTable->setColumnWidth(4, 100);   // æ“ä½œåˆ— - åŠ åˆ°100pxï¼ŒæŒ‰é’®æœ‰è¶³å¤Ÿç©ºé—´

    for (int i = 0; i < cabins.size(); ++i) {
        const Cabin &cabin = cabins[i];

        // è®¾ç½®åˆé€‚çš„è¡Œé«˜
        ui->cabinTable->setRowHeight(i, 48);  // å¢åŠ è¡Œé«˜ï¼Œè®©å†…å®¹æœ‰å‘¼å¸ç©ºé—´

        // èˆ±ä½ç±»å‹ - ä½¿ç”¨ç¨å¤§å­—ä½“
        QTableWidgetItem *typeItem = new QTableWidgetItem(cabin.getCabinType());
        typeItem->setTextAlignment(Qt::AlignCenter);
        typeItem->setFont(QFont("Microsoft YaHei", 11));  // å¢å¤§å­—ä½“
        ui->cabinTable->setItem(i, 0, typeItem);

        // ä»·æ ¼ - åªä¿ç•™æ•´æ•°ï¼Œä½†æ˜¾ç¤ºæ¸…æ™°
        double price = cabin.getPrice();
        QString priceText = QString("Â¥%1").arg(static_cast<int>(price));
        QTableWidgetItem *priceItem = new QTableWidgetItem(priceText);
        priceItem->setTextAlignment(Qt::AlignRight | Qt::AlignVCenter);
        priceItem->setForeground(QColor(220, 53, 69));
        priceItem->setFont(QFont("Microsoft YaHei", 12, QFont::Bold));  // å¢å¤§å¹¶åŠ ç²—
        ui->cabinTable->setItem(i, 1, priceItem);

        // è¡Œæé¢åº¦ - æ¸…æ™°æ˜¾ç¤º
        QTableWidgetItem *baggageItem = new QTableWidgetItem(cabin.getBaggageAllowance());
        baggageItem->setTextAlignment(Qt::AlignCenter);
        baggageItem->setFont(QFont("Microsoft YaHei", 11));  // å¢å¤§å­—ä½“
        ui->cabinTable->setItem(i, 2, baggageItem);

        // å¯ç”¨åº§ä½ - å®Œæ•´æ˜¾ç¤º
        int availableSeats = cabin.getAvailableSeats();
        int totalSeats = cabin.getTotalSeats();
        QString seatsText;

        if (availableSeats <= 0) {
            seatsText = "å·²å”®ç½„";
        } else if (availableSeats < 5) {
            seatsText = QString("ä»…å‰©%1åº§").arg(availableSeats);  // æ¢å¤å®Œæ•´æ–‡å­—
        } else {
            seatsText = QString("%1/%2").arg(availableSeats).arg(totalSeats);
        }

        QTableWidgetItem *seatsItem = new QTableWidgetItem(seatsText);
        seatsItem->setTextAlignment(Qt::AlignCenter);
        seatsItem->setFont(QFont("Microsoft YaHei", 11));

        // æ ¹æ®åº§ä½æƒ…å†µè®¾ç½®é¢œè‰²
        if (availableSeats <= 0) {
            seatsItem->setForeground(QColor(108, 117, 125));
        } else if (availableSeats < 5) {
            seatsItem->setForeground(QColor(220, 53, 69));
            seatsItem->setFont(QFont("Microsoft YaHei", 11, QFont::Bold));
        } else if (availableSeats < 10) {
            seatsItem->setForeground(QColor(255, 153, 0));
        } else {
            seatsItem->setForeground(QColor(40, 167, 69));
        }

        ui->cabinTable->setItem(i, 3, seatsItem);

        // é¢„è®¢æŒ‰é’® - é€‚å½“å¤§å°ï¼Œæœ‰è¶³å¤Ÿç©ºé—´
        QString buttonText = availableSeats > 0 ? "é¢„è®¢" : "å·²å”®ç½„";
        QPushButton *bookButton = new QPushButton(buttonText);

        // æŒ‰é’®å¤§å°é€‚ä¸­ï¼Œé€‚åˆ100pxåˆ—å®½
        bookButton->setFixedSize(85, 34);

        if (availableSeats > 0) {
            bookButton->setStyleSheet(
                "QPushButton {"
                "    background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #28a745, stop:1 #20c997);"
                "    color: white;"
                "    border: none;"
                "    border-radius: 4px;"   // é€‚å½“åœ†è§’
                "    font-size: 12px;"      // å¢å¤§å­—ä½“
                "    font-weight: bold;"
                "    font-family: 'Microsoft YaHei';"
                "}"
                "QPushButton:hover {"
                "    background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #20c997, stop:1 #17a2b8);"
                "}"
                "QPushButton:pressed {"
                "    background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #1e7e34, stop:1 #138496);"
                "}"
                );
            bookButton->setCursor(Qt::PointingHandCursor);
        } else {
            bookButton->setStyleSheet(
                "QPushButton {"
                "    background: #6c757d;"
                "    color: white;"
                "    border: none;"
                "    border-radius: 4px;"
                "    font-size: 12px;"      // å¢å¤§å­—ä½“
                "    font-weight: bold;"
                "    font-family: 'Microsoft YaHei';"
                "}"
                );
        }

        bookButton->setEnabled(availableSeats > 0);

        connect(bookButton, &QPushButton::clicked, this, [this, i]() {
            onBookButtonClicked(i);
        });

        ui->cabinTable->setCellWidget(i, 4, bookButton);
    }

    // è®¾ç½®è¡¨æ ¼æ ·å¼ - æ›´æ¸…æ™°çš„è¡¨å¤´
    ui->cabinTable->horizontalHeader()->setStyleSheet(
        "QHeaderView::section {"
        "    background-color: #f8f9fa;"
        "    padding: 10px 8px;"           // å¢åŠ å†…è¾¹è·
        "    border: 1px solid #dee2e6;"
        "    font-weight: bold;"
        "    font-size: 12px;"             // å¢å¤§è¡¨å¤´å­—ä½“
        "    font-family: 'Microsoft YaHei';"
        "    color: #495057;"
        "}"
        );

    // è®¾ç½®è¡¨æ ¼æ•´ä½“æ ·å¼
    ui->cabinTable->setStyleSheet(
        "QTableWidget {"
        "    gridline-color: #dee2e6;"
        "    background-color: white;"
        "    alternate-background-color: #f8f9fa;"  // éš”è¡Œå˜è‰²
        "    selection-background-color: #e3f2fd;"
        "}"
        "QTableWidget::item {"
        "    padding: 6px 8px;"            // å¢åŠ æ ¼å­å†…è¾¹è·
        "    border-bottom: 1px solid #f1f3f4;"
        "}"
        "QTableWidget::item:selected {"
        "    background-color: #e3f2fd;"
        "    color: #1e88e5;"
        "}"
        );

    // å¯ç”¨éš”è¡Œå˜è‰²
    ui->cabinTable->setAlternatingRowColors(true);
}

void FlightDetailDialog::onBookButtonClicked(int row)
{
    if (row >= 0 && row < currentCabins.size()) {
        const Cabin &cabin = currentCabins[row];
        showBookingDialog(cabin);
    }
}

void FlightDetailDialog::showBookingDialog(const Cabin &cabin)
{
    BookingDialog *dialog = new BookingDialog(flight, cabin, currentUsername,
                                              networkManager, this);
    dialog->setAttribute(Qt::WA_DeleteOnClose);
    dialog->exec();
}

void FlightDetailDialog::onCloseButtonClicked()
{
    this->accept();  // å…³é—­å¯¹è¯æ¡†
}

void FlightDetailDialog::onMessageReceived(const NetworkMessage &message)
{
    if (message.type == CABIN_SEARCH_RESPONSE) {
        bool success = message.data["success"].toBool();

        if (success) {
            QJsonArray cabinsArray = message.data["cabins"].toArray();
            QList<Cabin> cabins;

            for (auto it = cabinsArray.constBegin(); it != cabinsArray.constEnd(); ++it) {
                Cabin cabin = Cabin::fromJson((*it).toObject());
                cabins.append(cabin);
            }

            displayCabinData(cabins);
        } else {
            QMessageBox::warning(this, "é”™è¯¯", "åŠ è½½èˆ±ä½ä¿¡æ¯å¤±è´¥ï¼š" + message.data["message"].toString());
        }
    }
}
